Code/Framework/AzCore/AzCore/Serialization/Json/MapSerializer.cpp

```diff
@@ -333,6 +333,33 @@ namespace AZ
             {
                 // Convert the array to a Object with the member being the key.
                 outputValue.SetObject();
+
+                if (sortResult)
+                {
+                    // RapidJSON has a feature enabled called RAPIDJSON_USE_MEMBERSMAP.
+                    // When defined, RapidJSON system stores a multimap in addition to the array of members
+                    // to speed up lookups so that find and erase operations can be O(logn) instead of linear.
+                    // However, because of this, sorting the array of members without updating the map (which has
+                    // no API to do that) will break the object's data structure.
+                    // so to keep these objects in order, but not call sort() on the actual constructed
+                    // json document, instead, sort first, before inserting.
+                    auto less = [](const rapidjson::Value& lhs,
+                                   const rapidjson::Value& rhs) -> bool
+                    {
+                        JsonSerializerCompareResult result = JsonSerialization::Compare(lhs, rhs);
+                        if (result == JsonSerializerCompareResult::Equal)
+                        {
+                            return JsonSerialization::Compare(lhs, rhs) == JsonSerializerCompareResult::Less;
```

Comment:
Won't this always return false?

Reply:
yeah, this looks like a mistake on my part.  PReviously it would compare the keys, then the values.  Now it compares the object directly twice.  I could probably just reutrn Compare(lhs, rhs) == Less 

since I'm assuming that Compare will compare the elements of the object anyway...

Reply:
I think you can just do JsonSerialization::Compare(lhs, rhs) here. Depending on the type of data in the map (stored as an object vs stored as an array representing multimap), corresponding Compare() calls will be called accordingly. 
What was being done previously to compare keys first is also a redundant operations since they will be anyway compared if you pass along 2 maps.

Reply:
Yes, I agree.  Looking at the code here, the JsonSerialize::Compare will compare types of the objects first, if the types differ, will return the "smallest" type (according to the enum of possible types).  If the types are the same, it will compare the values itself, and it compares arrays by comparing them element-wise (ie, if they are an array of [key, value], its going to first compare the keys, then the values.

Reply:
The [JSONSerialization::CompareArray](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/Json/JsonSerialization.cpp#L522-L542) compares the array sizes first and if they differ it returns the smaller array as the one that is less than(which I don't know if it is correct).

That means that for two arrays
```
lhs = ["abcd"]
rhs = ["abc", 32]
```
The lhs side would be less than(<) the right hand side.

But that is inconsistent with the [JSONSerialization::CompareString](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/Json/JsonSerialization.cpp#L544-L554) which uses the string lexicographical compare which compares up to the minimum of the string characters sizes first.

If the strings differ at that point, it returns the result of that comparison.
If for some reason they are equal, only then does it compare the sizes.
The logic for lexicographical compare is in the string_view [compare](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/string/string_view.h#L784-L803) function.

So for a string of
```
lhs = "abcd"
rhs = "abc"
```

The left hand side would be greater than(>) the right hand side.

Reply:
Yeah, ultimately I think this would be okay to just use compare from jsonserialization.  The important purpose of sorting here is not to establish a specific order, but instead to establish a stable order so that things from an unordered map or unordered set don't randomize every time you save the file.

I'll switch it out to just compare the json objects directly.